das/aggregator.go

```diff
@@ -0,0 +1,218 @@
+			servicesThatSignedCert = append(servicesThatSignedCert, d)
+			pubKeys = append(pubKeys, d.pubKey)
+		}
+	}
+	signedBlob := serializeSignableFields(*requestedCert)
+	sigMatch, err := blsSignatures.VerifySignature(requestedCert.Sig, signedBlob, blsSignatures.AggregatePublicKeys(pubKeys))
+	if err != nil {
+		return nil, err
+	}
+	if !sigMatch {
+		return nil, errors.New("Signature of data in cert passed in doesn't match")
+	}
+
+	blobChan := make(chan []byte)
+	errorChan := make(chan error)
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for _, d := range servicesThatSignedCert {
+		go func(ctx context.Context, d serviceDetails) {
+			blob, err := d.service.Retrieve(ctx, cert)
+			if err != nil {
+				errorChan <- err
+				return
+			}
+			var blobHash [32]byte
+			copy(blobHash[:], crypto.Keccak256(blob))
+			if blobHash == requestedCert.DataHash {
+				blobChan <- blob
+			} else {
+				errorChan <- fmt.Errorf("DAS (mask %X) returned data that doesn't match requested hash!", d.signerMask)
+			}
+		}(subCtx, d)
+	}
+
+	errorCount := 0
+	var errorCollection []error
+	for errorCount < len(servicesThatSignedCert) {
+		select {
+		case blob := <-blobChan:
+			return blob, nil
+		case err = <-errorChan:
+			errorCollection = append(errorCollection, err)
+			log.Warn("Couldn't retrieve message from DAS", "err", err)
+			errorCount++
+		case <-ctx.Done():
+			break
+		}
+	}
+
+	return nil, fmt.Errorf("Data wasn't able to be retrieved from any DAS: %v", errorCollection)
+}
+
+type storeResponse struct {
+	cert *arbstate.DataAvailabilityCertificate
+	err  error
+
+	details serviceDetails
+}
+
+// Store calls Store on each backend DAS in parallel and collects responses; if
+// there were enough (at least K) responses then it aggregates the signatures and
+// signerMasks from each DAS together to put in the DataAvailabilityCertificate
+// that it returns; if it gets enough errors that K successes is impossible, then
+// it stops early and returns an error, and if it gets not enough successful
+// responses by the time its context is canceled (eg via DeadlineWrapper) then it
+// also returns an error.
+func (a *Aggregator) Store(ctx context.Context, message []byte, timeout uint64) (*arbstate.DataAvailabilityCertificate, error) {
+	var aggSignersMask uint64
+	var pubKeys []blsSignatures.PublicKey
+	var sigs []blsSignatures.Signature
+	var aggCert arbstate.DataAvailabilityCertificate
+
+	var initialStoreSucceeded bool
+	storeFailures := 0
+	if timeout == CALLEE_PICKS_TIMEOUT {
+		timeout = uint64(time.Now().Add(a.config.retentionPeriod).Unix())
+	}
+
+	responses := make(chan storeResponse)
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for _, d := range a.services {
+		go func(ctx context.Context, d serviceDetails) {
+			cert, err := d.service.Store(ctx, message, timeout)
+			responses <- storeResponse{cert, err, d}
+		}(subCtx, d)
+	}
+
+	successfullyStoredCount := 0
+	for i := 0; i < len(a.services); i++ {
```

Comment:
It seems better to count successes (as we already do), count failures, and break out of the loop as soon as we either have enough successes or too many failures.  Then the loop would be an infinite `for {` loop.

Reply:
This would cause us to accumulate just enough successful signatures to create a cert.  So there might be some DASes who are storing the data but aren't included in the cert. That seems fine -- it could happen anyway under the current structure, if some DAS signed but the context timed out before its signature was received.

Because of the possibility of storage without a signature in the cert, it makes sense in `Retrieve` to query every DAS, not just the ones whose signatures are in the cert.  If the number of required signatures is close to the total numbers of DASes (the usual case), this won't be much extra work, and it might allow us to retrieve data that would otherwise not be retrievable.

Reply:
I initially had it set up the way you are suggesting, ie return as soon as K signatures have been returned and cancel the remaining queries. I noticed that the final certificate then almost always had only K signatures. 

I decided to change it to wait until either the deadline was reached or all N responses  were gathered. My reasoning was if we are specifying a deadline for the call, that's the amount of time we are prepared to wait for the call to return, so why not get confirmations that we've stored the data as many times as possible during that time.

But your point is taken that we will likely be setting K to be close to N, so I will change the behavior of `Retrieve` now and we can discuss `Store` early termination some more today and come to a decision.